tools/system_libs.py

```diff
@@ -1638,6 +1652,10 @@ def add_library(lib):
   if shared.Settings.JS_MATH:
     libs_to_link = [(system_libs_map['libjsmath'].get_path(), True)] + libs_to_link
 
+  # to override the normal libc printf, we must come before it
+  if shared.Settings.PRINTF_LONG_DOUBLE:
+    libs_to_link = [(system_libs_map['libprintf_long_double'].get_path(), True)] + libs_to_link
```

Comment:
If you can, I would put this on line 1577 right before libc (using the same add_library helper)

Reply:
I guess that same goes for jsmath too?

Reply:
I don't follow? Line 1577 is
```python
        for haz in has_syms:
          if haz in need_syms:
            # remove symbols that are supplied by another of the inputs
            need_syms.remove(haz) # This is 1577..?
        if shared.Settings.VERBOSE:
          logger.debug('considering %s: we need %s and have %s' % (lib.name, str(need_syms), str(has_syms)))
```
So I'm not sure where you are suggesting to put this. But maybe the bigger issue is that this can't be moved earlier - it has to be after the `sort` operation on line 1641.

Reply:
Maybe you need to sync to head?

This line should come just before: `add_library(system_libs_map['libc'])`..  order is important for all of the `add_library` function here.

The sort operations are are only for specific cases.. they won't apply to libc or this library (I will try to remove them.. they are porbably not needed anymore since that most recent changes here).

Reply:
Thanks, done.

(Oh, I assumed when you said line numbers they were relative to this PR... otherwise which HEAD would I look at..? it can keep changing outside. Is there a convention I'm not aware of?)

Reply:
I just just been lazy to looking at my local line numbers...  Your are right its not a very accurate way to communicate file positions :)